### ISSUE TITLE
move cuda-version to 9.x

### ISSUE THREAD

**Issue by kwmonroe** [juju-solutions/layer-nvidia-cuda#17]
Latest cuda is now 9.1.85-1. We should update `cuda-version` in config.yaml to this new version.

This is not terribly urgent because the 8.x cuda-repo installs the latest cuda-9.x packages anyway, but it would be nice to make the config match what actually happens on the system.

Note: the 8.x cuda-repo is apparently signed nicely, so it installs without issue.  The 9.x repo package needs a key added to apt.  See the relevant parts of the nvidia readme:

http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#ubuntu-installation

```
$ sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/<distro>/<architecture>/7fa2af80.pub
```

If we do not add this key, install will fail with an insecure package error:

```
2018-03-19 14:16:59 DEBUG install WARNING: The following packages cannot be authenticated!
2018-03-19 14:16:59 DEBUG install   nvidia-390 nvidia-390-dev libcuda1-390 nvidia-modprobe libxnvctrl0
2018-03-19 14:16:59 DEBUG install   nvidia-settings nvidia-opencl-icd-390 cuda-drivers
2018-03-19 14:16:59 DEBUG install E: There were unauthenticated packages and -y was used without --allow-unauthenticated
2018-03-19 14:16:59 ERROR juju-log Hook error:
```

**Comment by kwmonroe** [comment]
Something else to note, I do not see mention of explicitly installing the `nvidia-3xx` drivers in the install guide:

http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#ubuntu-installation

Perhaps this means we can get rid of `function all:all:install_nvidia_driver()`:

https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/reactive/cuda.sh#L98

Whomever takes this issue should try manually doing the `apt install cuda` part to see if we `nvidia-3xx` packages automatically.  If so, we can drop that function (and all callers).

**Comment by ktsakalozos** [comment]
In terms of drivers, there is the cuda-drivers meta-package that should bring in the right nvidia drivers suited for cuda based on the repo you are in.

However, we said we would move the cuda-driver installation inside the docker layer since the docker layer can now install the nvidia docker and therefore it has the nvidia driver as a prerequisite. Right now the docker layer is doing things it shouldn't eg https://github.com/juju-solutions/layer-docker/blob/master/reactive/docker.py#L298

**Comment by kwmonroe** [comment]
> However, we said we would move the cuda-driver installation inside the docker layer

I think i'm missing something.. If the big `cuda` metapackage brings in `cuda-drivers`, what is left to move to layer-docker?  Won't those packages already be installed by this layer?

**Comment by kwmonroe** [comment]
Confirming that installing the big 'ol `cuda` metapackage does bring in `cuda-drivers`, which brings in `nvida-3xx`:

```bash
$ wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/cuda-repo-ubuntu1604_9.1.85-1_amd64.deb
$ sudo dpkg -i cuda-repo-ubuntu1604_9.1.85-1_amd64.deb
$ sudo apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/7fa2af80.pub
$ sudo apt-get update
$ sudo apt-get install cuda
.... coffeeeeeeeeee ....
$ dpkg -l | grep -E 'cuda-repo|cuda-driver|nvidia-'
ii  cuda-driver-dev-9-1                   9.1.85-1                                   amd64        CUDA Driver native dev stub library
ii  cuda-drivers                          390.30-1                                   amd64        CUDA Driver meta-package
ii  cuda-repo-ubuntu1604                  9.1.85-1                                   amd64        cuda repository configuration files
ii  nvidia-390                            390.30-0ubuntu1                            amd64        NVIDIA binary driver - version 390.30
ii  nvidia-390-dev                        390.30-0ubuntu1                            amd64        NVIDIA binary Xorg driver development files
ii  nvidia-modprobe                       390.30-0ubuntu1                            amd64        Load the NVIDIA kernel driver and create device files
ii  nvidia-opencl-icd-390                 390.30-0ubuntu1                            amd64        NVIDIA OpenCL ICD
ii  nvidia-prime                          0.8.2                                      amd64        Tools to enable NVIDIA's Prime
ii  nvidia-settings                       390.30-0ubuntu1                            amd64        Tool for configuring the NVIDIA graphics driver
```

So my previous question still stands.. What's moving to `layer-docker` if this layer handles the `nvidia-3xx` packages already?

**Comment by ktsakalozos** [comment]
The cuda-drivers meta-package is the only package needed. Everything else this layer is doing is not needed. As soon as we have the docker layer install cuda-drivers we will stop using this layer!

We need to move the cuda-drivers installation on the docker layer because  the nvidia-docker them to be there. We have to coordinate the drivers installation/removal with what the user is requesting to be installed on the docker layer.

**Comment by kwmonroe** [comment]
Ahhhhh, i see your true colors now @ktsakalozos.  K8s doesn't need this layer once layer-docker does the needful, but for everybody else using this layer (eg, bigtop), it's good to know that the big `cuda` package is all that's needed to get drivers installed.

So i'm back to suggesting we remove `function all:all:install_nvidia_driver()` function.

**Comment by kwmonroe** [comment]
The urgency just went up.  When we install `cuda-repo` 8.x and subsequently the `cuda` metapackage, a few things happen.

The `cuda` meta package brings in the latest debs from the nvidia repository, which includes `cuda-drivers`, which nicely puts everybody on 9.x.

However, *this* layer also forcefully installs the `nvidia-375` packages (probably from a time when `cuda-drivers` didn't exist):

https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/reactive/cuda.sh#L107

```
    apt-get install -yqq --no-install-recommends \
        nvidia-375 \
        nvidia-375-dev \
        libcuda1-375
```

Fortunately, nvdidia has made this a transitional package that facilitates the installation of `nvidia-390`.  Unfortuantely, this transition happens to install `nvidia-384` first.  Here's where it gets gross:

https://packages.ubuntu.com/xenial/misc/libcuda1-384

As you can see, libcuda1-384 has had a security update in the ubuntu archives.  If the timing is right,  we'll put down `nvidia-375`, which transitions to `nvidia-384`, which transitions to `nvidia-390`.  At that point, the OS unattended security upgrades will detect that -384 is out of date, upgrade it, and break -390.

You'll know when this has happened because `nvidia-smi` on the GPU machine will return something like "library/runtime mismatch" and you'll see an update to `nvidia-384` in `/var/log/apt/history.log`:

```
$ cat /var/log/apt/history.log |grep cuda | tail -2
Upgrade: libcuda1-375:amd64 (384.111-0ubuntu0.16.04.1, 390.30-0ubuntu1), nvidia-375-dev:amd64 (384.111-0ubuntu0.16.04.1, 390.30-0ubuntu1), nvidia-375:amd64 (384.111-0ubuntu0.16.04.1, 390.30-0ubuntu1)
Remove: nvidia-384-dev:amd64 (384.111-0ubuntu0.16.04.1), libcuda1-384:amd64 (384.111-0ubuntu0.16.04.1), nvidia-384:amd64 (384.111-0ubuntu0.16.04.1)
```

To fix it, we need to set the default `cuda-version` to 9.x, handle the apt key, and remove the explicit calls to install `nvidia-375`.  This will prevent `nvidia-384` from ever being on the system, hence preventing the security update that breaks `nvidia-390`.

**Comment by kwmonroe** [comment]
FYI, the reason 8.x does not have the package signing problem is because the 8.x repo deb does a wget and `apt-key add` in postinst:

```
$ cat postinst
#! /bin/sh

set -e

case "$1" in
    configure)
        #cat /var/cuda-repo/7fa2af80.pub | apt-key add - || echo "Failed to add GPGKEY at /var/cuda-repo/7fa2af80.pub to apt keys."
        wget -q -O - http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/7fa2af80.pub | apt-key add - || echo "Failed to add GPGKEY at http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/7fa2af80.pub to apt keys."
        ;;
esac
```

By contrast, the 9.x repo deb *tells* the user what to do (`cat <<EOM`), but doesn't actually do it:

```
$ cat postinst
#! /bin/sh

set -e

case "$1" in
    configure)
        keyexists="$(gpg /etc/apt/trusted.gpg 2>/dev/null | grep -i 7fa2af80)" || true
        if [ ! -f "/etc/apt/trusted.gpg" ] || [ ! "$keyexists" ]; then
            #addKey="sudo apt-key add /var/cuda-repo/7fa2af80.pub"
            addKey="sudo apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/7fa2af80.pub"
            cat <<EOM

The public CUDA GPG key does not appear to be installed.
To install the key, run this command:
$addKey

EOM
        fi
        ;;
esac
```

**Comment by kwmonroe** [comment]
Fixed with #18
